{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import seqpandas as spd\n",
    "from Bio import SeqIO\n",
    "import gzip\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Sample files used"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "gbk = \"tests/test-data/GCF_000013425.1.gbk.gz\"\n",
    "fasta = \"tests/test-data/random_sequences.fasta.gz\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Reading Different Genomic File Formats\n",
    "\n",
    "## FASTA Files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>_seq</th>\n",
       "      <th>id</th>\n",
       "      <th>name</th>\n",
       "      <th>description</th>\n",
       "      <th>dbxrefs</th>\n",
       "      <th>features</th>\n",
       "      <th>annotations</th>\n",
       "      <th>_per_letter_annotations</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>(C, G, A, T, A, T, T, C, G, A, T, C, C, G, C, ...</td>\n",
       "      <td>FWIRNKE01DKIF6</td>\n",
       "      <td>FWIRNKE01DKIF6</td>\n",
       "      <td>FWIRNKE01DKIF6 rank=0000177 x=1346.0 y=2772.0 ...</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>(G, C, G, A, G, C, A, G, C, A, A, T, C, A, T, ...</td>\n",
       "      <td>FWIRNKE01CDBE3</td>\n",
       "      <td>FWIRNKE01CDBE3</td>\n",
       "      <td>FWIRNKE01CDBE3 rank=0000320 x=854.0 y=2685.0 l...</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>(C, G, A, G, C, A, G, C, A, C, A, T, C, A, T, ...</td>\n",
       "      <td>FWIRNKE01BKZJJ</td>\n",
       "      <td>FWIRNKE01BKZJJ</td>\n",
       "      <td>FWIRNKE01BKZJJ rank=0000535 x=531.0 y=3933.0 l...</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                _seq              id  \\\n",
       "0  (C, G, A, T, A, T, T, C, G, A, T, C, C, G, C, ...  FWIRNKE01DKIF6   \n",
       "1  (G, C, G, A, G, C, A, G, C, A, A, T, C, A, T, ...  FWIRNKE01CDBE3   \n",
       "2  (C, G, A, G, C, A, G, C, A, C, A, T, C, A, T, ...  FWIRNKE01BKZJJ   \n",
       "\n",
       "             name                                        description dbxrefs  \\\n",
       "0  FWIRNKE01DKIF6  FWIRNKE01DKIF6 rank=0000177 x=1346.0 y=2772.0 ...      []   \n",
       "1  FWIRNKE01CDBE3  FWIRNKE01CDBE3 rank=0000320 x=854.0 y=2685.0 l...      []   \n",
       "2  FWIRNKE01BKZJJ  FWIRNKE01BKZJJ rank=0000535 x=531.0 y=3933.0 l...      []   \n",
       "\n",
       "  features annotations _per_letter_annotations  \n",
       "0       []          {}                    None  \n",
       "1       []          {}                    None  \n",
       "2       []          {}                    None  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fasta_df = spd.read_seq(fasta, format=\"fasta\")\n",
    "fasta_df.head(3)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# You can also import seqrecords directly from BioPythons parsing output."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>_seq</th>\n",
       "      <th>id</th>\n",
       "      <th>name</th>\n",
       "      <th>description</th>\n",
       "      <th>dbxrefs</th>\n",
       "      <th>features</th>\n",
       "      <th>annotations</th>\n",
       "      <th>_per_letter_annotations</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>(C, G, A, T, A, T, T, C, G, A, T, C, C, G, C, ...</td>\n",
       "      <td>FWIRNKE01DKIF6</td>\n",
       "      <td>FWIRNKE01DKIF6</td>\n",
       "      <td>FWIRNKE01DKIF6 rank=0000177 x=1346.0 y=2772.0 ...</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>(G, C, G, A, G, C, A, G, C, A, A, T, C, A, T, ...</td>\n",
       "      <td>FWIRNKE01CDBE3</td>\n",
       "      <td>FWIRNKE01CDBE3</td>\n",
       "      <td>FWIRNKE01CDBE3 rank=0000320 x=854.0 y=2685.0 l...</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>(C, G, A, G, C, A, G, C, A, C, A, T, C, A, T, ...</td>\n",
       "      <td>FWIRNKE01BKZJJ</td>\n",
       "      <td>FWIRNKE01BKZJJ</td>\n",
       "      <td>FWIRNKE01BKZJJ rank=0000535 x=531.0 y=3933.0 l...</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                _seq              id  \\\n",
       "0  (C, G, A, T, A, T, T, C, G, A, T, C, C, G, C, ...  FWIRNKE01DKIF6   \n",
       "1  (G, C, G, A, G, C, A, G, C, A, A, T, C, A, T, ...  FWIRNKE01CDBE3   \n",
       "2  (C, G, A, G, C, A, G, C, A, C, A, T, C, A, T, ...  FWIRNKE01BKZJJ   \n",
       "\n",
       "             name                                        description dbxrefs  \\\n",
       "0  FWIRNKE01DKIF6  FWIRNKE01DKIF6 rank=0000177 x=1346.0 y=2772.0 ...      []   \n",
       "1  FWIRNKE01CDBE3  FWIRNKE01CDBE3 rank=0000320 x=854.0 y=2685.0 l...      []   \n",
       "2  FWIRNKE01BKZJJ  FWIRNKE01BKZJJ rank=0000535 x=531.0 y=3933.0 l...      []   \n",
       "\n",
       "  features annotations _per_letter_annotations  \n",
       "0       []          {}                    None  \n",
       "1       []          {}                    None  \n",
       "2       []          {}                    None  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "with gzip.open(fasta, \"rt\") as handle:\n",
    "    seqrecords = SeqIO.parse(handle, format=\"fasta\")\n",
    "    fasta_df = spd.BioDataFrame.from_seqrecords(seqrecords)\n",
    "fasta_df.head(3)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Search description with little effort!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>_seq</th>\n",
       "      <th>id</th>\n",
       "      <th>name</th>\n",
       "      <th>description</th>\n",
       "      <th>dbxrefs</th>\n",
       "      <th>features</th>\n",
       "      <th>annotations</th>\n",
       "      <th>_per_letter_annotations</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>(C, G, A, T, A, T, T, C, G, A, T, C, C, G, C, ...</td>\n",
       "      <td>FWIRNKE01DKIF6</td>\n",
       "      <td>FWIRNKE01DKIF6</td>\n",
       "      <td>FWIRNKE01DKIF6 rank=0000177 x=1346.0 y=2772.0 ...</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>{}</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                _seq              id  \\\n",
       "0  (C, G, A, T, A, T, T, C, G, A, T, C, C, G, C, ...  FWIRNKE01DKIF6   \n",
       "\n",
       "             name                                        description dbxrefs  \\\n",
       "0  FWIRNKE01DKIF6  FWIRNKE01DKIF6 rank=0000177 x=1346.0 y=2772.0 ...      []   \n",
       "\n",
       "  features annotations _per_letter_annotations  \n",
       "0       []          {}                    None  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# I want all ranks of a certain number\n",
    "fasta_df[fasta_df[\"description\"].str.contains(\"rank=0000177\")]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Better view of the layout for a single row"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'_seq': Seq('CGATATTCGATCCGCATCGCTGCCCTACCCGTGGAGTGCCTCCCTCGGNGCAG'),\n",
       "  'id': 'FWIRNKE01DKIF6',\n",
       "  'name': 'FWIRNKE01DKIF6',\n",
       "  'description': 'FWIRNKE01DKIF6 rank=0000177 x=1346.0 y=2772.0 length=53',\n",
       "  'dbxrefs': [],\n",
       "  'features': [],\n",
       "  'annotations': {},\n",
       "  '_per_letter_annotations': None}]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fasta_df.head(1).to_dict(\"records\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## SAM/BAM (Sequence Alignment Map) Files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SAM shape: (5, 12)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>flag</th>\n",
       "      <th>ref_name</th>\n",
       "      <th>ref_pos</th>\n",
       "      <th>map_quality</th>\n",
       "      <th>cigar</th>\n",
       "      <th>next_ref_name</th>\n",
       "      <th>next_ref_pos</th>\n",
       "      <th>length</th>\n",
       "      <th>seq</th>\n",
       "      <th>qual</th>\n",
       "      <th>tags</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>r001</td>\n",
       "      <td>99</td>\n",
       "      <td>ref1</td>\n",
       "      <td>7</td>\n",
       "      <td>30</td>\n",
       "      <td>8M2I4M1D3M</td>\n",
       "      <td>=</td>\n",
       "      <td>37</td>\n",
       "      <td>39</td>\n",
       "      <td>TTAGATAAAGGATACTG</td>\n",
       "      <td>*</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>r002</td>\n",
       "      <td>0</td>\n",
       "      <td>ref1</td>\n",
       "      <td>9</td>\n",
       "      <td>30</td>\n",
       "      <td>3S6M1P1I4M</td>\n",
       "      <td>*</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>AAAAGATAAGGATA</td>\n",
       "      <td>*</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>r003</td>\n",
       "      <td>0</td>\n",
       "      <td>ref1</td>\n",
       "      <td>9</td>\n",
       "      <td>30</td>\n",
       "      <td>5S6M</td>\n",
       "      <td>*</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>GCCTAAGCTAA</td>\n",
       "      <td>*</td>\n",
       "      <td>[SA:Z:ref2,29,-,6H5M,17,0;]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>r004</td>\n",
       "      <td>0</td>\n",
       "      <td>ref1</td>\n",
       "      <td>16</td>\n",
       "      <td>30</td>\n",
       "      <td>6M14N5M</td>\n",
       "      <td>*</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>ATAGCTTCAGC</td>\n",
       "      <td>*</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>r003</td>\n",
       "      <td>2064</td>\n",
       "      <td>ref2</td>\n",
       "      <td>29</td>\n",
       "      <td>17</td>\n",
       "      <td>6H5M</td>\n",
       "      <td>*</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>TAGGC</td>\n",
       "      <td>*</td>\n",
       "      <td>[SA:Z:ref1,9,+,5S6M,30,1;]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   name  flag ref_name ref_pos map_quality       cigar next_ref_name  \\\n",
       "0  r001    99     ref1       7          30  8M2I4M1D3M             =   \n",
       "1  r002     0     ref1       9          30  3S6M1P1I4M             *   \n",
       "2  r003     0     ref1       9          30        5S6M             *   \n",
       "3  r004     0     ref1      16          30     6M14N5M             *   \n",
       "4  r003  2064     ref2      29          17        6H5M             *   \n",
       "\n",
       "  next_ref_pos length                seq qual                         tags  \n",
       "0           37     39  TTAGATAAAGGATACTG    *                           []  \n",
       "1            0      0     AAAAGATAAGGATA    *                           []  \n",
       "2            0      0        GCCTAAGCTAA    *  [SA:Z:ref2,29,-,6H5M,17,0;]  \n",
       "3            0      0        ATAGCTTCAGC    *                           []  \n",
       "4            0      0              TAGGC    *   [SA:Z:ref1,9,+,5S6M,30,1;]  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Read SAM file for sequence alignments\n",
    "sam_file = \"tests/test-data/example.sam\"\n",
    "sam_df = spd.read_seq(sam_file, format=\"sam\")\n",
    "print(f\"SAM shape: {sam_df.shape}\")\n",
    "sam_df.head()\n",
    "# Note: SAM reading requires pysam library and proper SAM formatting"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## BED (Browser Extensible Data) Files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "BED shape: (6, 12)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>chromStart</th>\n",
       "      <th>chromEnd</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>thickStart</th>\n",
       "      <th>thickEnd</th>\n",
       "      <th>itemRgb</th>\n",
       "      <th>blockCount</th>\n",
       "      <th>blockSizes</th>\n",
       "      <th>blockStarts</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1000</td>\n",
       "      <td>2000</td>\n",
       "      <td>feature1</td>\n",
       "      <td>100</td>\n",
       "      <td>+</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3000</td>\n",
       "      <td>4000</td>\n",
       "      <td>feature2</td>\n",
       "      <td>200</td>\n",
       "      <td>-</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr2</td>\n",
       "      <td>5000</td>\n",
       "      <td>6000</td>\n",
       "      <td>feature3</td>\n",
       "      <td>150</td>\n",
       "      <td>+</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr2</td>\n",
       "      <td>7000</td>\n",
       "      <td>8000</td>\n",
       "      <td>feature4</td>\n",
       "      <td>300</td>\n",
       "      <td>.</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr3</td>\n",
       "      <td>9000</td>\n",
       "      <td>10000</td>\n",
       "      <td>feature5</td>\n",
       "      <td>250</td>\n",
       "      <td>-</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chrom  chromStart  chromEnd      name  score strand thickStart thickEnd  \\\n",
       "0  chr1        1000      2000  feature1    100      +                       \n",
       "1  chr1        3000      4000  feature2    200      -                       \n",
       "2  chr2        5000      6000  feature3    150      +                       \n",
       "3  chr2        7000      8000  feature4    300      .                       \n",
       "4  chr3        9000     10000  feature5    250      -                       \n",
       "\n",
       "  itemRgb blockCount blockSizes blockStarts  \n",
       "0                                            \n",
       "1                                            \n",
       "2                                            \n",
       "3                                            \n",
       "4                                            "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Read BED file for genomic intervals\n",
    "bed_file = \"tests/test-data/example.bed\"\n",
    "bed_df = spd.read_bed(bed_file)\n",
    "print(f\"BED shape: {bed_df.shape}\")\n",
    "bed_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## VCF (Variant Call Format) Files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF shape: (6, 12)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>QUAL</th>\n",
       "      <th>FILTER</th>\n",
       "      <th>INFO</th>\n",
       "      <th>FORMAT</th>\n",
       "      <th>Sample1</th>\n",
       "      <th>Sample2</th>\n",
       "      <th>Sample3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>14370</td>\n",
       "      <td>rs6054257</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>29</td>\n",
       "      <td>PASS</td>\n",
       "      <td>NS=3;DP=14;AF=0.5;DB;H2</td>\n",
       "      <td>GT:GQ:DP:HQ</td>\n",
       "      <td>{'GT': '0|0', 'GQ': '48', 'DP': '1', 'HQ': '51...</td>\n",
       "      <td>{'GT': '1|0', 'GQ': '48', 'DP': '8', 'HQ': '51...</td>\n",
       "      <td>{'GT': '1/1', 'GQ': '43', 'DP': '5', 'HQ': '.,.'}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>17330</td>\n",
       "      <td>.</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>3</td>\n",
       "      <td>q10</td>\n",
       "      <td>NS=3;DP=11;AF=0.017</td>\n",
       "      <td>GT:GQ:DP:HQ</td>\n",
       "      <td>{'GT': '0|0', 'GQ': '49', 'DP': '3', 'HQ': '58...</td>\n",
       "      <td>{'GT': '0|1', 'GQ': '3', 'DP': '5', 'HQ': '65,3'}</td>\n",
       "      <td>{'GT': '0/0', 'GQ': '41', 'DP': '3', 'HQ': '.,.'}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>1110696</td>\n",
       "      <td>rs6040355</td>\n",
       "      <td>A</td>\n",
       "      <td>G,T</td>\n",
       "      <td>67</td>\n",
       "      <td>PASS</td>\n",
       "      <td>NS=2;DP=10;AF=0.333,0.667;AA=T;DB</td>\n",
       "      <td>GT:GQ:DP:HQ</td>\n",
       "      <td>{'GT': '1|2', 'GQ': '21', 'DP': '6', 'HQ': '23...</td>\n",
       "      <td>{'GT': '2|1', 'GQ': '2', 'DP': '0', 'HQ': '18,2'}</td>\n",
       "      <td>{'GT': '2/2', 'GQ': '35', 'DP': '4', 'HQ': '.,.'}</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  CHROM      POS         ID REF  ALT QUAL FILTER  \\\n",
       "0     1    14370  rs6054257   G    A   29   PASS   \n",
       "1     1    17330          .   T    A    3    q10   \n",
       "2     1  1110696  rs6040355   A  G,T   67   PASS   \n",
       "\n",
       "                                INFO       FORMAT  \\\n",
       "0            NS=3;DP=14;AF=0.5;DB;H2  GT:GQ:DP:HQ   \n",
       "1                NS=3;DP=11;AF=0.017  GT:GQ:DP:HQ   \n",
       "2  NS=2;DP=10;AF=0.333,0.667;AA=T;DB  GT:GQ:DP:HQ   \n",
       "\n",
       "                                             Sample1  \\\n",
       "0  {'GT': '0|0', 'GQ': '48', 'DP': '1', 'HQ': '51...   \n",
       "1  {'GT': '0|0', 'GQ': '49', 'DP': '3', 'HQ': '58...   \n",
       "2  {'GT': '1|2', 'GQ': '21', 'DP': '6', 'HQ': '23...   \n",
       "\n",
       "                                             Sample2  \\\n",
       "0  {'GT': '1|0', 'GQ': '48', 'DP': '8', 'HQ': '51...   \n",
       "1  {'GT': '0|1', 'GQ': '3', 'DP': '5', 'HQ': '65,3'}   \n",
       "2  {'GT': '2|1', 'GQ': '2', 'DP': '0', 'HQ': '18,2'}   \n",
       "\n",
       "                                             Sample3  \n",
       "0  {'GT': '1/1', 'GQ': '43', 'DP': '5', 'HQ': '.,.'}  \n",
       "1  {'GT': '0/0', 'GQ': '41', 'DP': '3', 'HQ': '.,.'}  \n",
       "2  {'GT': '2/2', 'GQ': '35', 'DP': '4', 'HQ': '.,.'}  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Read VCF file\n",
    "vcf_file = \"tests/test-data/vcf-test-data/example.vcf\"\n",
    "vcf_df = spd.read_vcf(vcf_file)\n",
    "print(f\"VCF shape: {vcf_df.shape}\")\n",
    "vcf_df.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sample1 genotype info for first variant:\n",
      "{'GT': '0|0', 'GQ': '48', 'DP': '1', 'HQ': '51,51'}\n"
     ]
    }
   ],
   "source": [
    "# VCF files contain variant information with sample genotypes\n",
    "# Sample columns are parsed as dictionaries containing GT, GQ, DP, etc.\n",
    "print(\"Sample1 genotype info for first variant:\")\n",
    "print(vcf_df[\"Sample1\"].iloc[0])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# GenBank to Pandas DataFrame example"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>_seq</th>\n",
       "      <th>id</th>\n",
       "      <th>name</th>\n",
       "      <th>description</th>\n",
       "      <th>dbxrefs</th>\n",
       "      <th>_per_letter_annotations</th>\n",
       "      <th>molecule_type</th>\n",
       "      <th>topology</th>\n",
       "      <th>data_file_division</th>\n",
       "      <th>date</th>\n",
       "      <th>accessions</th>\n",
       "      <th>sequence_version</th>\n",
       "      <th>keywords</th>\n",
       "      <th>source</th>\n",
       "      <th>organism</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>references</th>\n",
       "      <th>comment</th>\n",
       "      <th>contig</th>\n",
       "      <th>location</th>\n",
       "      <th>type</th>\n",
       "      <th>mol_type</th>\n",
       "      <th>strain</th>\n",
       "      <th>sub_species</th>\n",
       "      <th>db_xref</th>\n",
       "      <th>gene</th>\n",
       "      <th>locus_tag</th>\n",
       "      <th>note</th>\n",
       "      <th>codon_start</th>\n",
       "      <th>transl_table</th>\n",
       "      <th>product</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>translation</th>\n",
       "      <th>EC_number</th>\n",
       "      <th>pseudogene</th>\n",
       "      <th>gene_synonym</th>\n",
       "      <th>experiment</th>\n",
       "      <th>pseudo</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>(C, G, A, T, T, A, A, A, G, A, T, A, G, A, A, ...</td>\n",
       "      <td>NC_007795.1</td>\n",
       "      <td>NC_007795</td>\n",
       "      <td>Staphylococcus aureus subsp. aureus NCTC 8325 ...</td>\n",
       "      <td>[BioProject:PRJNA57795, Assembly:GCF_000013425.1]</td>\n",
       "      <td>{}</td>\n",
       "      <td>DNA</td>\n",
       "      <td>circular</td>\n",
       "      <td>CON</td>\n",
       "      <td>03-AUG-2016</td>\n",
       "      <td>[NC_007795]</td>\n",
       "      <td>1</td>\n",
       "      <td>[RefSeq]</td>\n",
       "      <td>Staphylococcus aureus subsp. aureus NCTC 8325</td>\n",
       "      <td>Staphylococcus aureus subsp. aureus NCTC 8325</td>\n",
       "      <td>[Bacteria, Firmicutes, Bacilli, Bacillales, St...</td>\n",
       "      <td>[location: [0:2821361]\\nauthors: Gillaspy,A.F....</td>\n",
       "      <td>REVIEWED REFSEQ: This record has been curated ...</td>\n",
       "      <td>join(CP000253.1:1..2821361)</td>\n",
       "      <td>(0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13,...</td>\n",
       "      <td>source</td>\n",
       "      <td>[genomic DNA]</td>\n",
       "      <td>[NCTC 8325]</td>\n",
       "      <td>[aureus]</td>\n",
       "      <td>[taxon:93061]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>(C, G, A, T, T, A, A, A, G, A, T, A, G, A, A, ...</td>\n",
       "      <td>NC_007795.1</td>\n",
       "      <td>NC_007795</td>\n",
       "      <td>Staphylococcus aureus subsp. aureus NCTC 8325 ...</td>\n",
       "      <td>[BioProject:PRJNA57795, Assembly:GCF_000013425.1]</td>\n",
       "      <td>{}</td>\n",
       "      <td>DNA</td>\n",
       "      <td>circular</td>\n",
       "      <td>CON</td>\n",
       "      <td>03-AUG-2016</td>\n",
       "      <td>[NC_007795]</td>\n",
       "      <td>1</td>\n",
       "      <td>[RefSeq]</td>\n",
       "      <td>Staphylococcus aureus subsp. aureus NCTC 8325</td>\n",
       "      <td>Staphylococcus aureus subsp. aureus NCTC 8325</td>\n",
       "      <td>[Bacteria, Firmicutes, Bacilli, Bacillales, St...</td>\n",
       "      <td>[location: [0:2821361]\\nauthors: Gillaspy,A.F....</td>\n",
       "      <td>REVIEWED REFSEQ: This record has been curated ...</td>\n",
       "      <td>join(CP000253.1:1..2821361)</td>\n",
       "      <td>(516, 517, 518, 519, 520, 521, 522, 523, 524, ...</td>\n",
       "      <td>gene</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[GeneID:3919798]</td>\n",
       "      <td>[dnaA]</td>\n",
       "      <td>[SAOUHSC_00001]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>(C, G, A, T, T, A, A, A, G, A, T, A, G, A, A, ...</td>\n",
       "      <td>NC_007795.1</td>\n",
       "      <td>NC_007795</td>\n",
       "      <td>Staphylococcus aureus subsp. aureus NCTC 8325 ...</td>\n",
       "      <td>[BioProject:PRJNA57795, Assembly:GCF_000013425.1]</td>\n",
       "      <td>{}</td>\n",
       "      <td>DNA</td>\n",
       "      <td>circular</td>\n",
       "      <td>CON</td>\n",
       "      <td>03-AUG-2016</td>\n",
       "      <td>[NC_007795]</td>\n",
       "      <td>1</td>\n",
       "      <td>[RefSeq]</td>\n",
       "      <td>Staphylococcus aureus subsp. aureus NCTC 8325</td>\n",
       "      <td>Staphylococcus aureus subsp. aureus NCTC 8325</td>\n",
       "      <td>[Bacteria, Firmicutes, Bacilli, Bacillales, St...</td>\n",
       "      <td>[location: [0:2821361]\\nauthors: Gillaspy,A.F....</td>\n",
       "      <td>REVIEWED REFSEQ: This record has been curated ...</td>\n",
       "      <td>join(CP000253.1:1..2821361)</td>\n",
       "      <td>(516, 517, 518, 519, 520, 521, 522, 523, 524, ...</td>\n",
       "      <td>CDS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[GeneID:3919798]</td>\n",
       "      <td>[dnaA]</td>\n",
       "      <td>[SAOUHSC_00001]</td>\n",
       "      <td>[binds to the dnaA-box as an ATP-bound complex...</td>\n",
       "      <td>[1]</td>\n",
       "      <td>[11]</td>\n",
       "      <td>[chromosomal replication initiation protein]</td>\n",
       "      <td>[YP_498609.1]</td>\n",
       "      <td>[MSEKEIWEKVLEIAQEKLSAVSYSTFLKDTELYTIKDGEAIVLSS...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                _seq           id       name  \\\n",
       "0  (C, G, A, T, T, A, A, A, G, A, T, A, G, A, A, ...  NC_007795.1  NC_007795   \n",
       "1  (C, G, A, T, T, A, A, A, G, A, T, A, G, A, A, ...  NC_007795.1  NC_007795   \n",
       "2  (C, G, A, T, T, A, A, A, G, A, T, A, G, A, A, ...  NC_007795.1  NC_007795   \n",
       "\n",
       "                                         description  \\\n",
       "0  Staphylococcus aureus subsp. aureus NCTC 8325 ...   \n",
       "1  Staphylococcus aureus subsp. aureus NCTC 8325 ...   \n",
       "2  Staphylococcus aureus subsp. aureus NCTC 8325 ...   \n",
       "\n",
       "                                             dbxrefs _per_letter_annotations  \\\n",
       "0  [BioProject:PRJNA57795, Assembly:GCF_000013425.1]                      {}   \n",
       "1  [BioProject:PRJNA57795, Assembly:GCF_000013425.1]                      {}   \n",
       "2  [BioProject:PRJNA57795, Assembly:GCF_000013425.1]                      {}   \n",
       "\n",
       "  molecule_type  topology data_file_division         date   accessions  \\\n",
       "0           DNA  circular                CON  03-AUG-2016  [NC_007795]   \n",
       "1           DNA  circular                CON  03-AUG-2016  [NC_007795]   \n",
       "2           DNA  circular                CON  03-AUG-2016  [NC_007795]   \n",
       "\n",
       "   sequence_version  keywords                                         source  \\\n",
       "0                 1  [RefSeq]  Staphylococcus aureus subsp. aureus NCTC 8325   \n",
       "1                 1  [RefSeq]  Staphylococcus aureus subsp. aureus NCTC 8325   \n",
       "2                 1  [RefSeq]  Staphylococcus aureus subsp. aureus NCTC 8325   \n",
       "\n",
       "                                        organism  \\\n",
       "0  Staphylococcus aureus subsp. aureus NCTC 8325   \n",
       "1  Staphylococcus aureus subsp. aureus NCTC 8325   \n",
       "2  Staphylococcus aureus subsp. aureus NCTC 8325   \n",
       "\n",
       "                                            taxonomy  \\\n",
       "0  [Bacteria, Firmicutes, Bacilli, Bacillales, St...   \n",
       "1  [Bacteria, Firmicutes, Bacilli, Bacillales, St...   \n",
       "2  [Bacteria, Firmicutes, Bacilli, Bacillales, St...   \n",
       "\n",
       "                                          references  \\\n",
       "0  [location: [0:2821361]\\nauthors: Gillaspy,A.F....   \n",
       "1  [location: [0:2821361]\\nauthors: Gillaspy,A.F....   \n",
       "2  [location: [0:2821361]\\nauthors: Gillaspy,A.F....   \n",
       "\n",
       "                                             comment  \\\n",
       "0  REVIEWED REFSEQ: This record has been curated ...   \n",
       "1  REVIEWED REFSEQ: This record has been curated ...   \n",
       "2  REVIEWED REFSEQ: This record has been curated ...   \n",
       "\n",
       "                        contig  \\\n",
       "0  join(CP000253.1:1..2821361)   \n",
       "1  join(CP000253.1:1..2821361)   \n",
       "2  join(CP000253.1:1..2821361)   \n",
       "\n",
       "                                            location    type       mol_type  \\\n",
       "0  (0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13,...  source  [genomic DNA]   \n",
       "1  (516, 517, 518, 519, 520, 521, 522, 523, 524, ...    gene            NaN   \n",
       "2  (516, 517, 518, 519, 520, 521, 522, 523, 524, ...     CDS            NaN   \n",
       "\n",
       "        strain sub_species           db_xref    gene        locus_tag  \\\n",
       "0  [NCTC 8325]    [aureus]     [taxon:93061]     NaN              NaN   \n",
       "1          NaN         NaN  [GeneID:3919798]  [dnaA]  [SAOUHSC_00001]   \n",
       "2          NaN         NaN  [GeneID:3919798]  [dnaA]  [SAOUHSC_00001]   \n",
       "\n",
       "                                                note codon_start transl_table  \\\n",
       "0                                                NaN         NaN          NaN   \n",
       "1                                                NaN         NaN          NaN   \n",
       "2  [binds to the dnaA-box as an ATP-bound complex...         [1]         [11]   \n",
       "\n",
       "                                        product     protein_id  \\\n",
       "0                                           NaN            NaN   \n",
       "1                                           NaN            NaN   \n",
       "2  [chromosomal replication initiation protein]  [YP_498609.1]   \n",
       "\n",
       "                                         translation EC_number pseudogene  \\\n",
       "0                                                NaN       NaN        NaN   \n",
       "1                                                NaN       NaN        NaN   \n",
       "2  [MSEKEIWEKVLEIAQEKLSAVSYSTFLKDTELYTIKDGEAIVLSS...       NaN        NaN   \n",
       "\n",
       "  gene_synonym experiment pseudo  \n",
       "0          NaN        NaN    NaN  \n",
       "1          NaN        NaN    NaN  \n",
       "2          NaN        NaN    NaN  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gbk_df = spd.read_seq(gbk, format=\"genbank\")\n",
    "gbk_df.head(3)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# See all the possible columns of meta data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['_seq', 'id', 'name', 'description', 'dbxrefs',\n",
       "       '_per_letter_annotations', 'molecule_type', 'topology',\n",
       "       'data_file_division', 'date', 'accessions', 'sequence_version',\n",
       "       'keywords', 'source', 'organism', 'taxonomy', 'references', 'comment',\n",
       "       'contig', 'location', 'type', 'mol_type', 'strain', 'sub_species',\n",
       "       'db_xref', 'gene', 'locus_tag', 'note', 'codon_start', 'transl_table',\n",
       "       'product', 'protein_id', 'translation', 'EC_number', 'pseudogene',\n",
       "       'gene_synonym', 'experiment', 'pseudo'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gbk_df.columns"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Seach for only rows of type CDS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Genome has CDS count: (2767, 38)\n",
      "Genome has GENE count: (2872, 38)\n"
     ]
    }
   ],
   "source": [
    "cds = gbk_df[gbk_df.type == \"CDS\"]\n",
    "gene = gbk_df[gbk_df.type == \"gene\"]\n",
    "print(\"Genome has CDS count:\", cds.shape)\n",
    "print(\"Genome has GENE count:\", gene.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>_seq</th>\n",
       "      <th>id</th>\n",
       "      <th>name</th>\n",
       "      <th>description</th>\n",
       "      <th>dbxrefs</th>\n",
       "      <th>_per_letter_annotations</th>\n",
       "      <th>molecule_type</th>\n",
       "      <th>topology</th>\n",
       "      <th>data_file_division</th>\n",
       "      <th>date</th>\n",
       "      <th>accessions</th>\n",
       "      <th>sequence_version</th>\n",
       "      <th>keywords</th>\n",
       "      <th>source</th>\n",
       "      <th>organism</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>references</th>\n",
       "      <th>comment</th>\n",
       "      <th>contig</th>\n",
       "      <th>location</th>\n",
       "      <th>type</th>\n",
       "      <th>mol_type</th>\n",
       "      <th>strain</th>\n",
       "      <th>sub_species</th>\n",
       "      <th>db_xref</th>\n",
       "      <th>gene</th>\n",
       "      <th>locus_tag</th>\n",
       "      <th>note</th>\n",
       "      <th>codon_start</th>\n",
       "      <th>transl_table</th>\n",
       "      <th>product</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>translation</th>\n",
       "      <th>EC_number</th>\n",
       "      <th>pseudogene</th>\n",
       "      <th>gene_synonym</th>\n",
       "      <th>experiment</th>\n",
       "      <th>pseudo</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>(C, G, A, T, T, A, A, A, G, A, T, A, G, A, A, ...</td>\n",
       "      <td>NC_007795.1</td>\n",
       "      <td>NC_007795</td>\n",
       "      <td>Staphylococcus aureus subsp. aureus NCTC 8325 ...</td>\n",
       "      <td>[BioProject:PRJNA57795, Assembly:GCF_000013425.1]</td>\n",
       "      <td>{}</td>\n",
       "      <td>DNA</td>\n",
       "      <td>circular</td>\n",
       "      <td>CON</td>\n",
       "      <td>03-AUG-2016</td>\n",
       "      <td>[NC_007795]</td>\n",
       "      <td>1</td>\n",
       "      <td>[RefSeq]</td>\n",
       "      <td>Staphylococcus aureus subsp. aureus NCTC 8325</td>\n",
       "      <td>Staphylococcus aureus subsp. aureus NCTC 8325</td>\n",
       "      <td>[Bacteria, Firmicutes, Bacilli, Bacillales, St...</td>\n",
       "      <td>[location: [0:2821361]\\nauthors: Gillaspy,A.F....</td>\n",
       "      <td>REVIEWED REFSEQ: This record has been curated ...</td>\n",
       "      <td>join(CP000253.1:1..2821361)</td>\n",
       "      <td>(516, 517, 518, 519, 520, 521, 522, 523, 524, ...</td>\n",
       "      <td>CDS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[GeneID:3919798]</td>\n",
       "      <td>[dnaA]</td>\n",
       "      <td>[SAOUHSC_00001]</td>\n",
       "      <td>[binds to the dnaA-box as an ATP-bound complex...</td>\n",
       "      <td>[1]</td>\n",
       "      <td>[11]</td>\n",
       "      <td>[chromosomal replication initiation protein]</td>\n",
       "      <td>[YP_498609.1]</td>\n",
       "      <td>[MSEKEIWEKVLEIAQEKLSAVSYSTFLKDTELYTIKDGEAIVLSS...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>(C, G, A, T, T, A, A, A, G, A, T, A, G, A, A, ...</td>\n",
       "      <td>NC_007795.1</td>\n",
       "      <td>NC_007795</td>\n",
       "      <td>Staphylococcus aureus subsp. aureus NCTC 8325 ...</td>\n",
       "      <td>[BioProject:PRJNA57795, Assembly:GCF_000013425.1]</td>\n",
       "      <td>{}</td>\n",
       "      <td>DNA</td>\n",
       "      <td>circular</td>\n",
       "      <td>CON</td>\n",
       "      <td>03-AUG-2016</td>\n",
       "      <td>[NC_007795]</td>\n",
       "      <td>1</td>\n",
       "      <td>[RefSeq]</td>\n",
       "      <td>Staphylococcus aureus subsp. aureus NCTC 8325</td>\n",
       "      <td>Staphylococcus aureus subsp. aureus NCTC 8325</td>\n",
       "      <td>[Bacteria, Firmicutes, Bacilli, Bacillales, St...</td>\n",
       "      <td>[location: [0:2821361]\\nauthors: Gillaspy,A.F....</td>\n",
       "      <td>REVIEWED REFSEQ: This record has been curated ...</td>\n",
       "      <td>join(CP000253.1:1..2821361)</td>\n",
       "      <td>(2155, 2156, 2157, 2158, 2159, 2160, 2161, 216...</td>\n",
       "      <td>CDS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[GeneID:3919799]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[SAOUHSC_00002]</td>\n",
       "      <td>[binds the polymerase to DNA and acts as a sli...</td>\n",
       "      <td>[1]</td>\n",
       "      <td>[11]</td>\n",
       "      <td>[DNA polymerase III subunit beta]</td>\n",
       "      <td>[YP_498610.1]</td>\n",
       "      <td>[MMEFTIKRDYFITQLNDTLKAISPRTTLPILTGIKIDAKEHEVIL...</td>\n",
       "      <td>[2.7.7.7]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>(C, G, A, T, T, A, A, A, G, A, T, A, G, A, A, ...</td>\n",
       "      <td>NC_007795.1</td>\n",
       "      <td>NC_007795</td>\n",
       "      <td>Staphylococcus aureus subsp. aureus NCTC 8325 ...</td>\n",
       "      <td>[BioProject:PRJNA57795, Assembly:GCF_000013425.1]</td>\n",
       "      <td>{}</td>\n",
       "      <td>DNA</td>\n",
       "      <td>circular</td>\n",
       "      <td>CON</td>\n",
       "      <td>03-AUG-2016</td>\n",
       "      <td>[NC_007795]</td>\n",
       "      <td>1</td>\n",
       "      <td>[RefSeq]</td>\n",
       "      <td>Staphylococcus aureus subsp. aureus NCTC 8325</td>\n",
       "      <td>Staphylococcus aureus subsp. aureus NCTC 8325</td>\n",
       "      <td>[Bacteria, Firmicutes, Bacilli, Bacillales, St...</td>\n",
       "      <td>[location: [0:2821361]\\nauthors: Gillaspy,A.F....</td>\n",
       "      <td>REVIEWED REFSEQ: This record has been curated ...</td>\n",
       "      <td>join(CP000253.1:1..2821361)</td>\n",
       "      <td>(3669, 3670, 3671, 3672, 3673, 3674, 3675, 367...</td>\n",
       "      <td>CDS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[GeneID:3919176]</td>\n",
       "      <td>NaN</td>\n",
       "      <td>[SAOUHSC_00003]</td>\n",
       "      <td>[conserved hypothetical protein]</td>\n",
       "      <td>[1]</td>\n",
       "      <td>[11]</td>\n",
       "      <td>[hypothetical protein]</td>\n",
       "      <td>[YP_498611.1]</td>\n",
       "      <td>[MIILVQEVVVEGDINLGQFLKTEGIIESGGQAKWFLQDVEVLING...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                _seq           id       name  \\\n",
       "2  (C, G, A, T, T, A, A, A, G, A, T, A, G, A, A, ...  NC_007795.1  NC_007795   \n",
       "4  (C, G, A, T, T, A, A, A, G, A, T, A, G, A, A, ...  NC_007795.1  NC_007795   \n",
       "6  (C, G, A, T, T, A, A, A, G, A, T, A, G, A, A, ...  NC_007795.1  NC_007795   \n",
       "\n",
       "                                         description  \\\n",
       "2  Staphylococcus aureus subsp. aureus NCTC 8325 ...   \n",
       "4  Staphylococcus aureus subsp. aureus NCTC 8325 ...   \n",
       "6  Staphylococcus aureus subsp. aureus NCTC 8325 ...   \n",
       "\n",
       "                                             dbxrefs _per_letter_annotations  \\\n",
       "2  [BioProject:PRJNA57795, Assembly:GCF_000013425.1]                      {}   \n",
       "4  [BioProject:PRJNA57795, Assembly:GCF_000013425.1]                      {}   \n",
       "6  [BioProject:PRJNA57795, Assembly:GCF_000013425.1]                      {}   \n",
       "\n",
       "  molecule_type  topology data_file_division         date   accessions  \\\n",
       "2           DNA  circular                CON  03-AUG-2016  [NC_007795]   \n",
       "4           DNA  circular                CON  03-AUG-2016  [NC_007795]   \n",
       "6           DNA  circular                CON  03-AUG-2016  [NC_007795]   \n",
       "\n",
       "   sequence_version  keywords                                         source  \\\n",
       "2                 1  [RefSeq]  Staphylococcus aureus subsp. aureus NCTC 8325   \n",
       "4                 1  [RefSeq]  Staphylococcus aureus subsp. aureus NCTC 8325   \n",
       "6                 1  [RefSeq]  Staphylococcus aureus subsp. aureus NCTC 8325   \n",
       "\n",
       "                                        organism  \\\n",
       "2  Staphylococcus aureus subsp. aureus NCTC 8325   \n",
       "4  Staphylococcus aureus subsp. aureus NCTC 8325   \n",
       "6  Staphylococcus aureus subsp. aureus NCTC 8325   \n",
       "\n",
       "                                            taxonomy  \\\n",
       "2  [Bacteria, Firmicutes, Bacilli, Bacillales, St...   \n",
       "4  [Bacteria, Firmicutes, Bacilli, Bacillales, St...   \n",
       "6  [Bacteria, Firmicutes, Bacilli, Bacillales, St...   \n",
       "\n",
       "                                          references  \\\n",
       "2  [location: [0:2821361]\\nauthors: Gillaspy,A.F....   \n",
       "4  [location: [0:2821361]\\nauthors: Gillaspy,A.F....   \n",
       "6  [location: [0:2821361]\\nauthors: Gillaspy,A.F....   \n",
       "\n",
       "                                             comment  \\\n",
       "2  REVIEWED REFSEQ: This record has been curated ...   \n",
       "4  REVIEWED REFSEQ: This record has been curated ...   \n",
       "6  REVIEWED REFSEQ: This record has been curated ...   \n",
       "\n",
       "                        contig  \\\n",
       "2  join(CP000253.1:1..2821361)   \n",
       "4  join(CP000253.1:1..2821361)   \n",
       "6  join(CP000253.1:1..2821361)   \n",
       "\n",
       "                                            location type mol_type strain  \\\n",
       "2  (516, 517, 518, 519, 520, 521, 522, 523, 524, ...  CDS      NaN    NaN   \n",
       "4  (2155, 2156, 2157, 2158, 2159, 2160, 2161, 216...  CDS      NaN    NaN   \n",
       "6  (3669, 3670, 3671, 3672, 3673, 3674, 3675, 367...  CDS      NaN    NaN   \n",
       "\n",
       "  sub_species           db_xref    gene        locus_tag  \\\n",
       "2         NaN  [GeneID:3919798]  [dnaA]  [SAOUHSC_00001]   \n",
       "4         NaN  [GeneID:3919799]     NaN  [SAOUHSC_00002]   \n",
       "6         NaN  [GeneID:3919176]     NaN  [SAOUHSC_00003]   \n",
       "\n",
       "                                                note codon_start transl_table  \\\n",
       "2  [binds to the dnaA-box as an ATP-bound complex...         [1]         [11]   \n",
       "4  [binds the polymerase to DNA and acts as a sli...         [1]         [11]   \n",
       "6                   [conserved hypothetical protein]         [1]         [11]   \n",
       "\n",
       "                                        product     protein_id  \\\n",
       "2  [chromosomal replication initiation protein]  [YP_498609.1]   \n",
       "4             [DNA polymerase III subunit beta]  [YP_498610.1]   \n",
       "6                        [hypothetical protein]  [YP_498611.1]   \n",
       "\n",
       "                                         translation  EC_number pseudogene  \\\n",
       "2  [MSEKEIWEKVLEIAQEKLSAVSYSTFLKDTELYTIKDGEAIVLSS...        NaN        NaN   \n",
       "4  [MMEFTIKRDYFITQLNDTLKAISPRTTLPILTGIKIDAKEHEVIL...  [2.7.7.7]        NaN   \n",
       "6  [MIILVQEVVVEGDINLGQFLKTEGIIESGGQAKWFLQDVEVLING...        NaN        NaN   \n",
       "\n",
       "  gene_synonym experiment pseudo  \n",
       "2          NaN        NaN    NaN  \n",
       "4          NaN        NaN    NaN  \n",
       "6          NaN        NaN    NaN  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cds.head(3)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Better view of the layout for a single row"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'_seq': Seq('CGATTAAAGATAGAAATACACGATGCGAGCAATCAAATTTCATAACATCACCAT...TAT'),\n",
       "  'id': 'NC_007795.1',\n",
       "  'name': 'NC_007795',\n",
       "  'description': 'Staphylococcus aureus subsp. aureus NCTC 8325 chromosome, complete genome',\n",
       "  'dbxrefs': ['BioProject:PRJNA57795', 'Assembly:GCF_000013425.1'],\n",
       "  '_per_letter_annotations': {},\n",
       "  'molecule_type': 'DNA',\n",
       "  'topology': 'circular',\n",
       "  'data_file_division': 'CON',\n",
       "  'date': '03-AUG-2016',\n",
       "  'accessions': ['NC_007795'],\n",
       "  'sequence_version': 1,\n",
       "  'keywords': ['RefSeq'],\n",
       "  'source': 'Staphylococcus aureus subsp. aureus NCTC 8325',\n",
       "  'organism': 'Staphylococcus aureus subsp. aureus NCTC 8325',\n",
       "  'taxonomy': ['Bacteria',\n",
       "   'Firmicutes',\n",
       "   'Bacilli',\n",
       "   'Bacillales',\n",
       "   'Staphylococcaceae',\n",
       "   'Staphylococcus'],\n",
       "  'references': [Reference(title='The Staphylococcus aureus NCTC8325 Genome', ...),\n",
       "   Reference(title='Direct Submission', ...),\n",
       "   Reference(title='Direct Submission', ...)],\n",
       "  'comment': 'REVIEWED REFSEQ: This record has been curated by NCBI staff. The\\nreference sequence was derived from CP000253.\\nRefSeq Category: Reference Genome\\n            UPR: UniProt Genome\\nStaphylococcus aureus subsp. aureus NCTC 8325 is available from\\nwww.narsa.net.\\nCOMPLETENESS: full length.',\n",
       "  'contig': 'join(CP000253.1:1..2821361)',\n",
       "  'location': SimpleLocation(ExactPosition(516), ExactPosition(1878), strand=1),\n",
       "  'type': 'CDS',\n",
       "  'mol_type': nan,\n",
       "  'strain': nan,\n",
       "  'sub_species': nan,\n",
       "  'db_xref': ['GeneID:3919798'],\n",
       "  'gene': ['dnaA'],\n",
       "  'locus_tag': ['SAOUHSC_00001'],\n",
       "  'note': ['binds to the dnaA-box as an ATP-bound complex at the origin of replication during the initiation of chromosomal replication; can also affect transcription of multiple genes including itself.'],\n",
       "  'codon_start': ['1'],\n",
       "  'transl_table': ['11'],\n",
       "  'product': ['chromosomal replication initiation protein'],\n",
       "  'protein_id': ['YP_498609.1'],\n",
       "  'translation': ['MSEKEIWEKVLEIAQEKLSAVSYSTFLKDTELYTIKDGEAIVLSSIPFNANWLNQQYAEIIQAILFDVVGYEVKPHFITTEELANYSNNETATPKETTKPSTETTEDNHVLGREQFNAHNTFDTFVIGPGNRFPHAASLAVAEAPAKAYNPLFIYGGVGLGKTHLMHAIGHHVLDNNPDAKVIYTSSEKFTNEFIKSIRDNEGEAFRERYRNIDVLLIDDIQFIQNKVQTQEEFFYTFNELHQNNKQIVISSDRPPKEIAQLEDRLRSRFEWGLIVDITPPDYETRMAILQKKIEEEKLDIPPEALNYIANQIQSNIRELEGALTRLLAYSQLLGKPITTELTAEALKDIIQAPKSKKITIQDIQKIVGQYYNVRIEDFSAKKRTKSIAYPRQIAMYLSRELTDFSLPKIGEEFGGRDHTTVIHAHEKISKDLKEDPIFKQEVENLEKEIRNV'],\n",
       "  'EC_number': nan,\n",
       "  'pseudogene': nan,\n",
       "  'gene_synonym': nan,\n",
       "  'experiment': nan,\n",
       "  'pseudo': nan}]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cds.head(1).to_dict(\"records\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "------"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Machine Learning Example Application"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### For simplicity of this example we will pretend we have a large sample set, but really we just using the single genbank genome. The follow will create the feature table we can use for a simple Sequential model.\n",
    "1. Sample the first 4 genes\n",
    "2. Given them random mutation value from 0 to 20\n",
    "3. Give each mutation a random \"observed in healthy patient\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['dnaA', 'recF', 'rplI', 'argJ']"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Sample for the first 4 genes\n",
    "genes = [gene[0] for gene in gbk_df[(gbk_df.type == \"gene\") & ~gbk_df.gene.isna()][\"gene\"][:4].values]\n",
    "genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dnaA</th>\n",
       "      <th>recF</th>\n",
       "      <th>rplI</th>\n",
       "      <th>argJ</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>19.926619</td>\n",
       "      <td>15.783713</td>\n",
       "      <td>17.041230</td>\n",
       "      <td>7.861119</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>7.200247</td>\n",
       "      <td>13.553658</td>\n",
       "      <td>15.470050</td>\n",
       "      <td>0.735451</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>13.831960</td>\n",
       "      <td>8.450581</td>\n",
       "      <td>7.379233</td>\n",
       "      <td>4.611984</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        dnaA       recF       rplI      argJ\n",
       "0  19.926619  15.783713  17.041230  7.861119\n",
       "1   7.200247  13.553658  15.470050  0.735451\n",
       "2  13.831960   8.450581   7.379233  4.611984"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Create fake observed mutation rates for target genomes compared to a reference genome\n",
    "mutation_array = lambda: list(np.random.uniform(low=0, high=20, size=(1, 100))[0])\n",
    "observed_gene_mutation_per_genome = spd.BioDataFrame.from_dict({gene: mutation_array() for gene in genes})\n",
    "observed_gene_mutation_per_genome.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dnaA</th>\n",
       "      <th>recF</th>\n",
       "      <th>rplI</th>\n",
       "      <th>argJ</th>\n",
       "      <th>fake_disease</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>19.926619</td>\n",
       "      <td>15.783713</td>\n",
       "      <td>17.041230</td>\n",
       "      <td>7.861119</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>7.200247</td>\n",
       "      <td>13.553658</td>\n",
       "      <td>15.470050</td>\n",
       "      <td>0.735451</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>13.831960</td>\n",
       "      <td>8.450581</td>\n",
       "      <td>7.379233</td>\n",
       "      <td>4.611984</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>6.591249</td>\n",
       "      <td>6.134858</td>\n",
       "      <td>12.047699</td>\n",
       "      <td>10.040812</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>9.614278</td>\n",
       "      <td>10.150649</td>\n",
       "      <td>15.374298</td>\n",
       "      <td>14.124760</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>16.860632</td>\n",
       "      <td>12.342782</td>\n",
       "      <td>10.999112</td>\n",
       "      <td>15.806463</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>0.620600</td>\n",
       "      <td>14.627543</td>\n",
       "      <td>4.200134</td>\n",
       "      <td>14.739281</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>6.340793</td>\n",
       "      <td>19.087236</td>\n",
       "      <td>8.459715</td>\n",
       "      <td>9.066288</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>8.825462</td>\n",
       "      <td>12.633047</td>\n",
       "      <td>7.450039</td>\n",
       "      <td>2.275944</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>19.892919</td>\n",
       "      <td>4.359332</td>\n",
       "      <td>18.910079</td>\n",
       "      <td>2.968047</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        dnaA       recF       rplI       argJ  fake_disease\n",
       "0  19.926619  15.783713  17.041230   7.861119           1.0\n",
       "1   7.200247  13.553658  15.470050   0.735451           0.0\n",
       "2  13.831960   8.450581   7.379233   4.611984           0.0\n",
       "3   6.591249   6.134858  12.047699  10.040812           0.0\n",
       "4   9.614278  10.150649  15.374298  14.124760           0.0\n",
       "5  16.860632  12.342782  10.999112  15.806463           0.0\n",
       "6   0.620600  14.627543   4.200134  14.739281           0.0\n",
       "7   6.340793  19.087236   8.459715   9.066288           0.0\n",
       "8   8.825462  12.633047   7.450039   2.275944           0.0\n",
       "9  19.892919   4.359332  18.910079   2.968047           0.0"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Give a 0 or 1 to show if a patient has the fake disease\n",
    "# We a biased 30 percent with disease\n",
    "patients = np.zeros(100)\n",
    "patients[:30] = 1\n",
    "np.random.shuffle(patients)\n",
    "\n",
    "observed_gene_mutation_per_genome[\"fake_disease\"] = list(patients)\n",
    "observed_gene_mutation_per_genome.head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "--------"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "seqpandas",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
